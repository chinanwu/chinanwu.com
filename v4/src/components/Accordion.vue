<template>
  <div>
    <button
      id="accordionBtn"
      class="Accordion__btn"
      @click="handleClick"
      @keydown="handleKeyDown"
    >
      <span>{{ label }}</span>
      <span>{{ expand ? "-" : "+" }}</span>
    </button>
    <transition name="expand" mode="out-in">
      <slot v-if="expand" name="content"></slot>
    </transition>
  </div>
</template>

<script>
import validEvent from "@/functions/validEvent";

export default {
  name: "Accordion",
  props: {
    label: {
      type: String,
      default: null
    },
    expand: {
      type: Boolean,
      default: false
    }
  },
  emits: ["change"],
  methods: {
    handleClick() {
      this.$emit("change");
    },
    handleKeyDown(event) {
      if (validEvent(event, ["Enter", "Space"])) {
        event.preventDefault();
        this.handleClick();
      }
    }
  }
};
</script>

<style lang="less" scoped>
@import "../styling/colours";

.Accordion__btn {
  width: 100%;
  text-align: left;
  padding: 1rem;
  background-color: @grey-05;
  border: none;
  display: flex;
  justify-content: space-between;
  cursor: pointer;
  transition: background-color 0.2s;
  border-radius: 0.5rem;
  font-size: 1.8rem;

  &:hover,
  &:focus {
    background-color: @grey-06;
  }

  &:active {
    background-color: @grey-04;
  }
}

.expand {
  &-enter-active,
  &-leave-active {
    transition: all 0.3s;
    overflow: hidden;
  }
  &-enter-from,
  &-leave-to {
    max-height: 0;
    opacity: 0;
  }
  &-enter-to,
  &-leave-from {
    max-height: 100vh;
    opacity: 1;
  }
}
</style>
