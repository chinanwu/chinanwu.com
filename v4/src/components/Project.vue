<template>
  <section class="Project" aria-labelledby="projectHeader">
    <Brackets>
      <template #content>
        <section class="Project__content" v-if="name !== 'SHOWCASE'">
          <header>
            <h3 class="headerLabel" id="projectHeader">{{ details.name }}</h3>
          </header>

          <div class="Project__details">
            <div class="Project__btns">
              <button class="Project__btn">
                <a :href="details.github">GITHUB</a>
              </button>
              <button
                class="Project__btn"
                :aria-disabled="!details.link"
                :aria-label="!details.link ? 'Live link coming soon' : null"
              >
                <a
                  :class="{ Project__linkTba: !details.link }"
                  :href="details.link"
                >
                  LINK
                </a>
              </button>
            </div>

            <p class="Project__description">
              {{ details.description }}
            </p>

            <section
              aria-labelledby="projectTechnologies"
              v-if="details.technologies"
            >
              <h4 id="projectTechnologies" class="Project__sectionHeader">
                Technologies
              </h4>
              <p
                class="Project__technology"
                v-if="details.technologies.backend"
              >
                <strong>Backend</strong>: {{ details.technologies.backend }}
              </p>
              <p
                class="Project__technology"
                v-if="details.technologies.frontend"
              >
                <strong>Frontend</strong>: {{ details.technologies.frontend }}
              </p>
            </section>

            <Accordion
              class="Project__accordion"
              :expand="expand"
              :label="expand ? 'No More Spice' : 'The Spicy Stuff'"
              @change="expand = !expand"
              v-if="details.features || details.faq"
            >
              <template #content>
                <section
                  :aria-label="'Extra information about ' + details.name"
                >
                  <section
                    aria-labelledby="projectFeatures"
                    v-if="details.features"
                  >
                    <h4 id="projectFeatures" class="Project__sectionHeader">
                      Features:
                    </h4>
                    <ul class="Project__features">
                      <li
                        class="Project__feature"
                        v-for="(feature, index) in details.features"
                        :key="name + '-feature-' + index"
                      >
                        {{ feature }}
                      </li>
                    </ul>
                  </section>

                  <section aria-labelledby="projectFaq" v-if="details.faq">
                    <h4 id="projectFaq" class="Project__sectionHeader">
                      FAQ
                    </h4>
                    <ol class="Project__qas">
                      <li
                        class="Project__qa"
                        v-for="(qa, index) in details.faq"
                        :key="name + '-faq-' + index"
                      >
                        <strong>{{ qa.question }}</strong
                        >: {{ qa.answer }}
                      </li>
                    </ol>
                  </section>
                </section>
              </template>
            </Accordion>
          </div>
        </section>
        <div v-else>
          <section aria-labelledby="showcaseHeader">
            <header>
              <h3 id="showcaseHeader" class="headerLabel">GITHUB SHOWCASE</h3>
            </header>

            <div class="Project__showcaseProjects">
              <section
                v-for="(project, index) in details.list"
                :key="'showcase-' + index"
                class="Project__showcase"
                :aria-labelledby="'showcaseHeader-' + index"
              >
                <header>
                  <h4 :id="'showcaseHeader-' + index" class="Project__header">
                    {{ project.name }}
                  </h4>
                </header>

                <p class="Project__descriptionShowcase">
                  {{ project.description }}
                </p>
                <p class="Project__descriptionShowcase">
                  Made with: <strong>{{ project.madeWith }}</strong>
                </p>
                <div class="Project__btns">
                  <button class="Project__btn">
                    <a :href="project.github">GITHUB</a>
                  </button>
                  <button class="Project__btn">
                    <a :href="project.link">
                      LINK
                    </a>
                  </button>
                </div>
              </section>
            </div>
          </section>
        </div>
      </template>
    </Brackets>
  </section>
</template>

<script>
import projects from "@/assets/data/projects";
import Accordion from "@/components/Accordion";
import Brackets from "@/components/Brackets";

export default {
  name: "Project",
  components: { Accordion, Brackets },
  props: {
    name: {
      type: String,
      required: true
    }
  },
  data() {
    return {
      details: {},
      expand: false
    };
  },
  methods: {
    getDetails() {
      projects.forEach(project => {
        if (project.key === this.name) {
          this.details = project;
        }
      });
    }
  },
  created() {
    this.getDetails();
  },
  updated() {
    this.getDetails();
  }
};
</script>

<style lang="less" scoped>
@import "../styling/colours.less";

.Project {
  display: flex;
  justify-content: center;
}

.Project__content {
  display: flex;
  flex-direction: column;
  width: 80%;
}

.Project__description,
.Project__descriptionShowcase {
  font-size: 1.8rem;
  text-align: center;
}

.Project__description {
  border-radius: 1rem;
  padding: 2rem;
  background: rgba(202, 254, 0, 0.25);

  transition: box-shadow 0.3s;
  &:hover {
    box-shadow: 0 0 0 0.8rem rgba(202, 254, 0, 0.1);
  }
}

.Project__btns {
  display: flex;
  justify-content: space-evenly;
  margin: 1rem 0;
}

.Project__btn {
  font-size: 2.4rem;
}

.Project__linkTba {
  color: @grey-03;
  cursor: not-allowed;

  &::after {
    content: "TBA";
    display: inline-block;
    transform: rotate(45deg) translateY(2rem) translateX(-3rem);
    background-color: @grey-01;
    color: @grey-06;
    font-size: 1.6rem;
  }
}

.Project__details--responsive {
  display: flex;
}

.Project__sectionHeader {
  text-align: center;
  font-size: 2.4rem;
  text-decoration: underline;
}

.Project__technology {
  font-size: 1.8rem;
  padding: 0;
  margin: 1rem 0;
}

.Project__accordion {
  margin: 2rem 0;
}

.Project__features,
.Project__qas {
  font-size: 1.8rem;
}

.Project__features {
  list-style: initial;
  padding-left: 3rem;
  margin-top: 1rem;
}

.Project__feature,
.Project__qa {
  margin: 0.5rem 0;
}

.Project__showcaseProjects {
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(300px, 1fr));
  grid-column-gap: 2rem;
}

.Project__showcase {
  background: rgba(202, 254, 0, 0.25);
  margin: 1rem 0;
  border-radius: 1rem;
  padding: 1rem 0;
}

.Project__header {
  font-size: 3rem;
  text-align: center;
  padding: 0;
}

@media only screen and (max-width: 625px) {
  .Project__content {
    width: initial;
  }
}
@media only screen and (max-width: 625px) {
  .Project__showcaseProjects {
    display: initial;
  }

  .Project__showcase {
    margin: 2rem 0;
  }
}

@media only screen and (max-width: 320px) {
  .Project__header {
    font-size: 2.4rem;
  }
}

@media only screen and (max-width: 250px) {
  .Project__btn {
    font-size: 1.8rem;
  }
}
</style>
