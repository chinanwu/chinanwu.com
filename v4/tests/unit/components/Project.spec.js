import Accordion from "@/components/Accordion";
import { mount } from "@vue/test-utils";
import Project from "@/components/Project.vue";

describe("Project", () => {
  describe("rendering", () => {
    it("renders", () => {
      const wrapper = mount(Project, {
        props: {
          name: "Test"
        },
        data() {
          return {
            details: {
              name: "Test",
              github: "Github",
              link: "Link",
              description: "test description",
              technologies: {
                frontend: "",
                backend: ""
              }
            }
          };
        }
      });
      expect(wrapper.element).toMatchSnapshot();
    });

    it("renders with a TBA link", () => {
      const wrapper = mount(Project, {
        props: {
          name: "Test"
        },
        data() {
          return {
            details: {
              name: "Test",
              github: "Github",
              description: "test description",
              technologies: {
                frontend: "",
                backend: ""
              }
            }
          };
        }
      });
      expect(wrapper.element).toMatchSnapshot();
    });

    it("renders with showcase", () => {
      const wrapper = mount(Project, {
        props: {
          name: "showcase"
        },
        data() {
          return {
            details: {
              name: "Showcase",
              github: "Github",
              description: "test description",
              technologies: {
                frontend: "",
                backend: ""
              }
            }
          };
        }
      });
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe("behaviour", () => {
    it("expands accordion with features", () => {
      const wrapper = mount(Project, {
        props: {
          name: "Test"
        },
        data() {
          return {
            details: {
              name: "Test",
              github: "Github",
              link: "Link",
              description: "test description",
              technologies: {
                frontend: "",
                backend: ""
              },
              features: ["list", "of", "features"]
            }
          };
        }
      });
      expect(wrapper.element).toMatchSnapshot();
      wrapper.findComponent(Accordion).vm.$emit("change");
      expect(wrapper.vm.expand).toBeTruthy();
    });
  });
});
