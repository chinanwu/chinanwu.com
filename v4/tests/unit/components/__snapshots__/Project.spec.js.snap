// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project behaviour expands accordion with features 1`] = `
<section
  aria-labelledby="projectHeader"
  class="Project"
>
  <div
    class="Brackets"
  >
    <div
      class="Bracket Bracket__upper"
      role="presentation"
    />
    
    <section
      class="Project__content"
    >
      <header>
        <h3
          class="headerLabel"
          id="projectHeader"
        >
          Test
        </h3>
      </header>
      <div
        class="Project__details"
      >
        <div
          class="Project__links"
        >
          <a
            aria-label="Test Github page"
            class="Project__link --uppercase"
            href="Github"
            title="Test Github page"
          >
             Github 
          </a>
          <a
            aria-disabled="false"
            aria-label="Test live project page"
            class="Project__link --uppercase"
            href="Link"
            title="Test live project page"
          >
             Link 
          </a>
        </div>
        <p
          class="Project__description"
        >
          test description
        </p>
        <section
          aria-labelledby="projectTechnologies"
        >
          <h4
            class="Project__sectionHeader"
            id="projectTechnologies"
          >
             Technologies 
          </h4>
          <!--v-if-->
          <!--v-if-->
        </section>
        <div
          class="Project__accordion"
        >
          <button
            aria-live="polite"
            class="Accordion__btn"
            id="accordionBtn"
          >
            <span>
              The Spicy Stuff
            </span>
            <span
              role="presentation"
            >
              +
            </span>
          </button>
          <transition-stub>
            <!--v-if-->
          </transition-stub>
        </div>
      </div>
    </section>
    
    <div
      class="Bracket Bracket__lower"
      role="presentation"
    />
  </div>
</section>
`;

exports[`Project rendering renders 1`] = `
<section
  aria-labelledby="projectHeader"
  class="Project"
>
  <div
    class="Brackets"
  >
    <div
      class="Bracket Bracket__upper"
      role="presentation"
    />
    
    <section
      class="Project__content"
    >
      <header>
        <h3
          class="headerLabel"
          id="projectHeader"
        >
          Test
        </h3>
      </header>
      <div
        class="Project__details"
      >
        <div
          class="Project__links"
        >
          <a
            aria-label="Test Github page"
            class="Project__link --uppercase"
            href="Github"
            title="Test Github page"
          >
             Github 
          </a>
          <a
            aria-disabled="false"
            aria-label="Test live project page"
            class="Project__link --uppercase"
            href="Link"
            title="Test live project page"
          >
             Link 
          </a>
        </div>
        <p
          class="Project__description"
        >
          test description
        </p>
        <section
          aria-labelledby="projectTechnologies"
        >
          <h4
            class="Project__sectionHeader"
            id="projectTechnologies"
          >
             Technologies 
          </h4>
          <!--v-if-->
          <!--v-if-->
        </section>
        <!--v-if-->
      </div>
    </section>
    
    <div
      class="Bracket Bracket__lower"
      role="presentation"
    />
  </div>
</section>
`;

exports[`Project rendering renders with a TBA link 1`] = `
<section
  aria-labelledby="projectHeader"
  class="Project"
>
  <div
    class="Brackets"
  >
    <div
      class="Bracket Bracket__upper"
      role="presentation"
    />
    
    <section
      class="Project__content"
    >
      <header>
        <h3
          class="headerLabel"
          id="projectHeader"
        >
          Test
        </h3>
      </header>
      <div
        class="Project__details"
      >
        <div
          class="Project__links"
        >
          <a
            aria-label="Test Github page"
            class="Project__link --uppercase"
            href="Github"
            title="Test Github page"
          >
             Github 
          </a>
          <a
            aria-disabled="true"
            aria-label="Test live project page - Coming soon"
            class="Project__linkTba Project__link --uppercase"
            title="Test live project page - Coming soon"
          >
             Link 
          </a>
        </div>
        <p
          class="Project__description"
        >
          test description
        </p>
        <section
          aria-labelledby="projectTechnologies"
        >
          <h4
            class="Project__sectionHeader"
            id="projectTechnologies"
          >
             Technologies 
          </h4>
          <!--v-if-->
          <!--v-if-->
        </section>
        <!--v-if-->
      </div>
    </section>
    
    <div
      class="Bracket Bracket__lower"
      role="presentation"
    />
  </div>
</section>
`;

exports[`Project rendering renders with showcase 1`] = `
<section
  aria-labelledby="projectHeader"
  class="Project"
>
  <div
    class="Brackets"
  >
    <div
      class="Bracket Bracket__upper"
      role="presentation"
    />
    
    <div>
      <section
        aria-labelledby="showcaseHeader"
      >
        <header>
          <h3
            class="headerLabel"
            id="showcaseHeader"
          >
             Github Showcase 
          </h3>
        </header>
        <div
          class="Project__showcaseProjects"
        >
          
          <section
            aria-labelledby="showcaseHeader-0"
            class="Project__showcase"
          >
            <header>
              <h4
                class="Project__header --uppercase"
                id="showcaseHeader-0"
              >
                Affirmer
              </h4>
            </header>
            <p
              class="Project__descriptionShowcase"
            >
              A site to made to help users with daily afirmations.
            </p>
            <p
              class="Project__descriptionShowcase"
            >
               Made with: 
              <strong>
                Vue
              </strong>
            </p>
            <div
              class="Project__links"
            >
              <a
                class="Project__link --uppercase"
                href="https://github.com/chinanwu/affirmer"
                title="Affirmer Github page"
              >
                 Github 
              </a>
              <a
                class="Project__link --uppercase"
                href="https://chinanwu.github.io/affirmer"
                title="Affirmer live project page"
              >
                 Link 
              </a>
            </div>
          </section>
          <section
            aria-labelledby="showcaseHeader-1"
            class="Project__showcase"
          >
            <header>
              <h4
                class="Project__header --uppercase"
                id="showcaseHeader-1"
              >
                Magnet-Poetry
              </h4>
            </header>
            <p
              class="Project__descriptionShowcase"
            >
              A fake fridge to make real magnetic poetry with!
            </p>
            <p
              class="Project__descriptionShowcase"
            >
               Made with: 
              <strong>
                Vue
              </strong>
            </p>
            <div
              class="Project__links"
            >
              <a
                class="Project__link --uppercase"
                href="https://github.com/chinanwu/magnet-poetry"
                title="Magnet-Poetry Github page"
              >
                 Github 
              </a>
              <a
                class="Project__link --uppercase"
                href="https://chinanwu.github.io/magnet-poetry"
                title="Magnet-Poetry live project page"
              >
                 Link 
              </a>
            </div>
          </section>
          <section
            aria-labelledby="showcaseHeader-2"
            class="Project__showcase"
          >
            <header>
              <h4
                class="Project__header --uppercase"
                id="showcaseHeader-2"
              >
                Solver
              </h4>
            </header>
            <p
              class="Project__descriptionShowcase"
            >
              A small go package that finds the optimal solution to Delta problems!
            </p>
            <p
              class="Project__descriptionShowcase"
            >
               Made with: 
              <strong>
                Golang
              </strong>
            </p>
            <div
              class="Project__links"
            >
              <a
                class="Project__link --uppercase"
                href="https://github.com/chinanwu/solver"
                title="Solver Github page"
              >
                 Github 
              </a>
              <a
                class="Project__link --uppercase"
                href="https://chinanwu.github.io/solver"
                title="Solver live project page"
              >
                 Link 
              </a>
            </div>
          </section>
          
        </div>
      </section>
    </div>
    
    <div
      class="Bracket Bracket__lower"
      role="presentation"
    />
  </div>
</section>
`;
